File: 206-reverse-linked-list.c
/* 翻转链表 双指针*/
#include <stdio.h>
#include <stdlib.h>

typedef struct ListNode {
    int val;
    struct ListNode *next;
} ListNode;

struct ListNode* reverseList(struct ListNode *head){
    struct ListNode *pre = NULL;
    struct ListNode *cur = head;

    while (cur != NULL) {
        struct ListNode *temp = cur->next;
        cur->next = pre;
        pre = cur;
        cur = temp;
    }
    return pre;
}

int main()
{
    // 创建链表
    ListNode node_5 = {5, NULL};
    ListNode node_4 = {4, &node_5};
    ListNode node_3 = {3, &node_4};
    ListNode node_2 = {2, &node_3};
    ListNode node_1 = {1, &node_2};

    // linked list node traverse
    ListNode *node_ptr = &node_1; // 指向头结点的指针
    while(node_ptr != NULL) {
        printf("%d\n", node_ptr->val);
        node_ptr = node_ptr->next;
    }

    ListNode *new_node = reverseList(&node_1);

    printf("\n");

    ListNode *node_ptr_2 = new_node; // 指向头结点的指针
    while(node_ptr_2 != NULL) {
        printf("%d\n", node_ptr_2->val);
        node_ptr_2 = node_ptr_2->next;
    }

    system("Pause");
    return 0;
}

File: 206-reverse-linked-list.py
# 206 翻转链表
# https://leetcode.cn/problems/reverse-linked-list/

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# def reverseList(head):
#     if not head:
#         return head
#     elif not head.next:
#         return head
#     else:
#         cur = head.next
#     pre = ListNode(head.val, next=None)  # 一个空节点

#     while cur.next:

#         tmp = cur.next
#         cur.next = pre
#         pre = cur
#         cur = tmp

#     # 循环过后，到达了最后一个节点
#     cur.next = pre
#     pre = cur

#     return pre

# 力扣上的例子
def reverseList(head):
    cur = head
    pre = None
    while cur:
        temp = cur.next
        cur.next = pre
        pre = cur
        cur = temp
    return pre

# 反转链表的递归方法 1
def reverse(pre, cur):
    """
    input type: list node
    """
    if not cur:
        return pre
    temp = cur.next
    cur.next = pre
    return reverse(cur, temp)

def reverseList_rec(head):
    return reverse(None, head)

#---------------------------------
# test
def traverse_linked_list(head):
    test = head
    while test:
        print(test.val)
        test = test.next

node_5 = ListNode(5)
node_4 = ListNode(4, node_5)
node_3 = ListNode(3, node_4)
node_2 = ListNode(2, node_3)
node_1 = ListNode(1, node_2)

only_one_node = ListNode(10)

traverse_linked_list(node_1)

newhead = reverseList(node_1)

traverse_linked_list(newhead)
